#Exercise 4.9 - Part 1

In [31]:
# Import liabraries
import pandas as pd
import numpy as np
import os

In [32]:
#Project folder path as a string
path = r'/Users/Nimrat/Documents/Instacart Basket Analysis'

In [33]:
#Import the “customers” data set into your notebook using the os library
df_custs = pd.read_csv(os.path.join(path, 'Original data', 'customers.csv'), index_col = False)

In [34]:
#To check the output
df_custs.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [35]:
#To check shape of the dataframe
df_custs.shape

(206209, 10)

In [36]:
#To describe the dataframe
df_custs.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

#4 - Wrangle the data so that it follows consistent logic; for example, rename columns with illogical names and drop columns that don’t add anything to your analysis.

In [37]:
#Fixing column names
df_custs.rename(columns = {'First Name': 'first_name', 'Surnam': 'last_name', 'Gender': 'gender', 'STATE': 'state', 'Age': 'age', 'n_dependants': 'n_dependents'}, inplace = True)

In [38]:
#To check output
df_custs.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependents fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

#5 - Complete the fundamental data quality and consistency checks you’ve learned throughout this Achievement; for example, check for and address missing values and duplicates, and convert any mixed-type data

In [39]:
#To find missing values in dataframe
df_custs.isnull().sum()

user_id             0
first_name      11259
last_name           0
gender              0
state               0
age                 0
date_joined         0
n_dependents        0
fam_status          0
income              0
dtype: int64

There are 11259 missing values in first_name. Since we don't use first and last name in the analysis so either we can drop these columns or kept as missing values.

In [40]:
#Finding duplicates
df_dups = df_custs[df_custs.duplicated()]

In [41]:
df_dups

Empty DataFrame
Columns: [user_id, first_name, last_name, gender, state, age, date_joined, n_dependents, fam_status, income]
Index: []

No duplicates were found

In [42]:
#Checking for mixed type data
df_custs.dtypes

user_id          int64
first_name      object
last_name       object
gender          object
state           object
age              int64
date_joined     object
n_dependents     int64
fam_status      object
income           int64
dtype: object

#6 - Combine your customer data with the rest of your prepared Instacart data. (Hint: Make sure the key columns are the same data type!)

In [43]:
#Import the “ords_prods_merge” data set into your notebook using the os library
ords_prods_merge = pd.read_pickle(os.path.join(path, 'Prepared data', 'ords_prods_merged_grouped.pkl'))

In [44]:
ords_prods_merge.shape

(32404859, 24)

In [45]:
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   ...    price_range_loc     busiest_day    busiest_days  \
0  ...  Mid-range product  Regularly busy  Regularly busy   
1  ...  Mid-range product  Regularly busy  Regularly busy   
2  ...  Mid-range product     Busiest day    Busiest days   
3  ...  Mid-range product  Regularly busy    Slowest days   
4  ...  Mid-range product      Least busy    Slowest days   

  busiest_period_of_day max_order      loyalty_flag avg_product_price  \
0           Most orders        32  Regular customer          6.935811   
1        Average orders        32  Regular customer          6.935811   
2        Average orders         5      New customer          7.930208   
3           Most orders         3      New customer          4.972414   
4        Average orders         3      New customer          4.972414   

  spending_flag  order_regularity_median order_frequency_flag  
0   Low spender                      8.0    Frequent customer  
1   Low spender                      8.0    Frequent customer  
2   Low spender                      8.0    Frequent customer  
3   Low spender                      9.0    Frequent customer  
4   Low spender                      9.0    Frequent customer  

[5 rows x 24 columns]

In [46]:
#Combining customers with ords_prods_merge
df_ords_prods_custs_merge = pd.merge(ords_prods_merge, df_custs, on='user_id', how='inner')

In [48]:
#Checking output
df_ords_prods_custs_merge.shape

(32404859, 33)

In [49]:
df_ords_prods_custs_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   ...  order_frequency_flag  first_name  last_name  gender      state age  \
0  ...     Frequent customer     Charles        Cox    Male  Minnesota  81   
1  ...     Frequent customer     Charles        Cox    Male  Minnesota  81   
2  ...     Frequent customer     Deborah      Glass  Female    Vermont  66   
3  ...     Frequent customer     Heather      Myers  Female  Wisconsin  40   
4  ...     Frequent customer     Heather      Myers  Female  Wisconsin  40   

  date_joined n_dependents  fam_status  income  
0    8/1/2019            1     married   49620  
1    8/1/2019            1     married   49620  
2   6/16/2018            2     married  158302  
3    2/9/2020            3     married   31308  
4    2/9/2020            3     married   31308  

[5 rows x 33 columns]

In [51]:
#To export dataframe
df_ords_prods_custs_merge.to_pickle(os.path.join(path, 'Prepared Data', 'ords_prods_custs_merged.pkl'))